src/mlpack/methods/cf/cf_main.cpp

```diff
@@ -523,23 +251,174 @@ static void mlpackMain()
+    else if (normalizationType == "item_mean")
+      cf->NormalizationType() = CFModel::ITEM_MEAN_NORMALIZATION;
+    else if (normalizationType == "user_mean")
+      cf->NormalizationType() = CFModel::USER_MEAN_NORMALIZATION;
+    else if (normalizationType == "overall_mean")
+      cf->NormalizationType() = CFModel::OVERALL_MEAN_NORMALIZATION;
+    else if (normalizationType == "z_score")
+      cf->NormalizationType() = CFModel::Z_SCORE_NORMALIZATION;
+
+    cf->Train(dataset,
+              neighborhood,
+              rank,
+              size_t(IO::GetParam<int>("max_iterations")),
+              IO::GetParam<double>("min_residue"),
+              IO::HasParam("iteration_only_termination"));
   }
   else
   {
     // Load from a model after validating parameters.
-    RequireAtLeastOnePassed({ "query", "all_user_recommendations",
-        "test" }, true);
+    RequireAtLeastOnePassed({ "query", "all_user_recommendations", "test" },
+        true);
 
     // Load an input model.
-    CFModel* c = std::move(IO::GetParam<CFModel*>("input_model"));
+    cf = std::move(IO::GetParam<CFModel*>("input_model"));
+  }
 
-    PerformAction(c);
+  // Get the types of the neighbor search method and the interpolation.  (These
+  // may or may not be used.)
+  NeighborSearchTypes nsType;
+  RequireParamInSet<string>("neighbor_search", { "cosine",
+      "euclidean", "pearson" }, true, "unknown neighbor search algorithm");
+  if (IO::GetParam<std::string>("neighbor_search") == "cosine")
+    nsType = COSINE_SEARCH;
+  else if (IO::GetParam<std::string>("neighbor_search") == "euclidean")
+    nsType = EUCLIDEAN_SEARCH;
+  else // if (IO::GetParam<std::string>("neighbor_search") == "pearson")
+    nsType = PEARSON_SEARCH;
+
+  InterpolationTypes interpolationType;
+  RequireParamInSet<string>("interpolation", { "average",
+      "regression", "similarity" }, true, "unknown interpolation algorithm");
+  if (IO::GetParam<std::string>("interpolation") == "average")
+    interpolationType = AVERAGE_INTERPOLATION;
+  else if (IO::GetParam<std::string>("interpolation") == "regression")
+    interpolationType = REGRESSION_INTERPOLATION;
+  else // if (IO::GetParam<std::string>("interpolation") == "similarity")
+    interpolationType = SIMILARITY_INTERPOLATION;
+
+  if (IO::HasParam("query") || IO::HasParam("all_user_recommendations"))
+  {
+    // Get parameters for generating recommendations.
+    const size_t numRecs = (size_t) IO::GetParam<int>("recommendations");
+
+    // Get the recommendations.
+    arma::Mat<size_t> recommendations;
+
+    // Reading users.
+    if (IO::HasParam("query"))
+    {
+      // User matrix.
+      arma::Mat<size_t> users =
+          std::move(IO::GetParam<arma::Mat<size_t>>("query"));
+      if (users.n_rows > 1)
+        users = users.t();
+      if (users.n_rows > 1)
+        Log::Fatal << "List of query users must be one-dimensional!"
+                   << std::endl;
```

Comment:
Maybe it makes sense to use `{}` here to improve readability, also in like 379-380 we use 4 spaces instead of aligning the `<<`.

Reply:
Agreed---fixed in 2f504f566. :+1: